### PR TITLE
Debugger: make skip-breaks depend on current code

### PR DESCRIPTION
**Pull request by grammati**
Previously, the var `*skip-breaks*` was storing a mode (:deeper,
:before, etc.) and a sequence of coordinates. However, it did not store
any information indicating what code the coordinates referred to.

This would lead to strange behavior when an :out or :here operation
would skip over code that called another instrumented function. The
debugger would treat the coordinates in *skip-breaks* as if they
referred to the code of the called function, and would seem (to the
user) to skip randomly over some or all of the function.

The fix:

1) Store *skip-breaks* as a map containing :mode, :coor, and :code,
where :code is the string of code that the :out or :here operation was
performed on.

2) Add logic to `skip-breaks?` to only skip if the to-be-skipped code
and the currently-being-debugged code are the same.

### CODE REVIEW

**Comment by Malabarba** [comment]
Thanks @grammati. This is something I've ran into a couple of times too.

The only thing that makes me unsure about this fix is the fact that assumes that each `defn` was instrumented in a separate `*msg*`.
Right now, this is always true, but that's actually something I consider a bug in the current implementation. (Currently, debugger just fails to instrument multiple functions in the same message, such as with `load-buffer` or `eval-region`).

That said, unless I get off my ass and fix the implementation itself, I can't really think of a better way to fix the `:here/:out` bug than the one you propose.

File: src/cider/nrepl/middleware/debug.clj [hunk 1]
```diff
@@ -102,21 +116,30 @@
...
     true))
 
 (defn skip-breaks!
   "Set the value of *skip-breaks* for the top-level breakpoint.
-  If bool-or-vec is a vector, mode should be :deeper or :before (see
-  `skip-breaks?`). Otherwise mode should be :all or false."
-  [mode & [bool-or-vec]]
-  (reset! *skip-breaks* (when mode
-                          (cons mode bool-or-vec))))
+  Additionl arguments depend on mode, and should be:
```
Comment (Malabarba) [inline src/cider/nrepl/middleware/debug.clj]: additionAl

File: src/cider/nrepl/middleware/debug.clj [hunk 2]
```diff
@@ -435,14 +445,13 @@
       (try
         (read-string code)
         (catch Exception e)))
-    (binding [*skip-breaks* (atom nil)]
```
Comment (Malabarba) [inline src/cider/nrepl/middleware/debug.clj]: I agree with this change, but PRs are easier to read if unrelated changes are left in separate commits. ;-)

File: src/cider/nrepl/middleware/debug.clj (hunk 1 above)
Comment (grammati) [inline src/cider/nrepl/middleware/debug.clj]: Thanks, clearly I wasn't using flyspell :)

File: src/cider/nrepl/middleware/debug.clj (hunk 2 above)
Comment (grammati) [inline src/cider/nrepl/middleware/debug.clj]: Fair point. I guess I considered this to be somewhat related, in that I made sure to carefully examine all uses of `*skip-breaks*`, and saw that this one was a leftover. Happy to move it to a separate PR if you prefer.

Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj]: Separate commit in the same PR would be fine.

Comment (grammati) [inline src/cider/nrepl/middleware/debug.clj]: OK, I put this part in a separate commit, rebased, and pushed.

**Comment by Malabarba** [comment]
Thanks again. I'll merge this as soon as I get a chance to read it again.
